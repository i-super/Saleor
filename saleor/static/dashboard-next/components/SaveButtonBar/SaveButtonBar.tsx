import Button from "@material-ui/core/Button";
import CircularProgress from "@material-ui/core/CircularProgress";
import gray from "@material-ui/core/colors/grey";
import { withStyles } from "@material-ui/core/styles";
import CheckIcon from "@material-ui/icons/Check";
import * as React from "react";
import i18n from "../../i18n";

export type SaveButtonBarState =
  | "loading"
  | "success"
  | "error"
  | "default"
  | string;
interface SaveButtonBarProps {
  disabled?: boolean;
  state?: SaveButtonBarState;
  labels?: {
    cancel?: string;
    delete?: string;
    save?: string;
  };
  onCancel: () => void;
  onDelete?: () => void;
  onSave(event: any);
}

const decorate = withStyles(theme => ({
  button: {
    marginRight: theme.spacing.unit
  },
  buttonProgress: {
    "& svg": {
      margin: 0
    },
    color: theme.palette.primary.main,
    marginBottom: -theme.spacing.unit * 1.5,
    marginLeft: -theme.spacing.unit * 0.5,
    marginRight: theme.spacing.unit * 0.5,
    marginTop: -theme.spacing.unit * 1.5,
    position: "relative" as "relative"
  },
  cancelButton: {
    marginRight: theme.spacing.unit * 2
  },
  deleteButton: {
    "&:hover": {
      backgroundColor: theme.palette.error.dark
    },
    backgroundColor: theme.palette.error.main,
    color: theme.palette.error.contrastText
  },
  error: {
    "&:hover": {
      backgroundColor: theme.palette.error.main
    },
    backgroundColor: theme.palette.error.main,
    color: theme.palette.error.contrastText
  },
  icon: {
    marginBottom: -theme.spacing.unit * 1.5,
    marginRight: theme.spacing.unit * 0.5,
    marginTop: -theme.spacing.unit * 1.5,
    position: "relative" as "relative"
  },
  root: {
    borderTop: `1px ${gray[300]} solid`,
    display: "flex",
    marginBottom: theme.spacing.unit * 2,
    marginTop: theme.spacing.unit * 2,
    paddingTop: theme.spacing.unit * 2,
    [theme.breakpoints.down("sm")]: {
      marginTop: theme.spacing.unit
    }
  },
  spacer: {
    flex: "1"
  },
  success: {
    "&:hover": {
      backgroundColor: theme.palette.primary.main
    },
    backgroundColor: theme.palette.primary.main,
    color: theme.palette.primary.contrastText
  }
}));
export const SaveButtonBar = decorate<SaveButtonBarProps>(
  ({
    classes,
    disabled,
    labels,
    state,
    onCancel,
    onDelete,
    onSave,
    ...props
  }) => {
    let buttonClassName;
    let buttonLabel;
    switch (state) {
      case "success":
        buttonClassName = classes.success;
        buttonLabel = i18n.t("Saved");
        break;
      case "error":
        buttonClassName = classes.error;
        buttonLabel = i18n.t("Error");
        break;
      case "loading":
        buttonClassName = "";
        buttonLabel = i18n.t("Saving");
        break;
      default:
        buttonClassName = "";
        buttonLabel = labels && labels.save ? labels.save : i18n.t("Save");
        break;
    }
    return (
      <div className={classes.root} {...props}>
        {!!onDelete && (
          <Button
            variant="contained"
            onClick={onDelete}
            className={classes.deleteButton}
          >
            {labels && labels.delete ? labels.delete : i18n.t("Remove")}
          </Button>
        )}
        <div className={classes.spacer} />
        <Button
          className={classes.cancelButton}
          variant="flat"
          onClick={onCancel}
        >
          {labels && labels.cancel ? labels.cancel : i18n.t("Cancel")}
        </Button>
        <Button
          variant="contained"
          onClick={onSave}
          color="secondary"
          disabled={disabled || state === "loading"}
          className={buttonClassName}
        >
          {state === "loading" && (
            <CircularProgress
              size={24}
              color="secondary"
              className={classes.buttonProgress}
            />
          )}
          {state === "success" && <CheckIcon className={classes.icon} />}
          {buttonLabel}
        </Button>
      </div>
    );
  }
);
SaveButtonBar.displayName = "SaveButtonBar";
export default SaveButtonBar;
