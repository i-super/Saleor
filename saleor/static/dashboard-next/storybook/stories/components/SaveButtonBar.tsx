import { storiesOf } from "@storybook/react";
import * as React from "react";

import SaveButtonBar from "../../../components/SaveButtonBar";
import Decorator from "../../Decorator";

const callbacks = {
  onCancel: undefined,
  onDelete: undefined,
  onSave: undefined
};

interface InteractiveDemoProps {
  targetState: "success" | "error" | "string";
}
interface InteractiveDemoState {
  state: "success" | "error" | "loading" | "default";
}

class InteractiveDemo extends React.Component<
  InteractiveDemoProps,
  InteractiveDemoState
> {
  state = {
    state: "default" as "default"
  };
  timer = undefined;

  componentWillUnmount() {
    clearTimeout(this.timer);
  }

  handleButtonClick = () => {
    if ((this.state.state as string) !== "loading") {
      this.setState(
        {
          state: "loading"
        },
        () => {
          this.timer = setTimeout(() => {
            this.setState({
              state: this.props.targetState as "success" | "error"
            });
          }, 2000);
        }
      );
    }
  };

  render() {
    return (
      <SaveButtonBar
        onSave={this.handleButtonClick}
        state={this.state.state as any}
      />
    );
  }
}

storiesOf("Generics / SaveButtonBar", module)
  .addDecorator(Decorator)
  .add("idle", () => <SaveButtonBar {...callbacks} />)
  .add("loading", () => <SaveButtonBar {...callbacks} state="loading" />)
  .add("success", () => <SaveButtonBar {...callbacks} state="success" />)
  .add("error", () => <SaveButtonBar {...callbacks} state="error" />)
  .add("disabled", () => <SaveButtonBar {...callbacks} disabled />)
  .add("interactive success", () => <InteractiveDemo targetState="success" />)
  .add("interactive error", () => <InteractiveDemo targetState="error" />);
