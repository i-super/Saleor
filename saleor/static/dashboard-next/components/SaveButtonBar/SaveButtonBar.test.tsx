import * as React from "react";
import * as renderer from "react-test-renderer";

import SaveButtonBar from "./SaveButtonBar";

describe("<SaveButtonBar />", () => {
  it("renders", () => {
    const component = renderer.create(
      <SaveButtonBar onBack={jest.fn()} onSave={jest.fn()} />
    );
    expect(component).toMatchSnapshot();
  });
  it("renders when disabled", () => {
    const component = renderer.create(
      <SaveButtonBar onBack={jest.fn()} onSave={jest.fn()} disabled />
    );
    expect(component).toMatchSnapshot();
  });
});
