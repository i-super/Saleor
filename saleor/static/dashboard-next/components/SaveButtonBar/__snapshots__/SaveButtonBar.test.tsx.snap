// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SaveButtonBar /> renders 1`] = `
<div
  className="Component-root-1"
>
  <div
    className="Component-spacer-2"
  />
  <button
    className="MuiButtonBase-root-19 MuiButton-root-4 Component-button-3"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-5"
    >
      Cancel
    </span>
    <span
      className="MuiTouchRipple-root-21"
    />
  </button>
  <button
    className="MuiButtonBase-root-19 MuiButton-root-4 MuiButton-raised-9 MuiButton-raisedPrimary-10"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-5"
    >
      Save
    </span>
    <span
      className="MuiTouchRipple-root-21"
    />
  </button>
</div>
`;

exports[`<SaveButtonBar /> renders when disabled 1`] = `
<div
  className="Component-root-1"
>
  <div
    className="Component-spacer-2"
  />
  <button
    className="MuiButtonBase-root-19 MuiButton-root-4 Component-button-3"
    disabled={false}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="0"
    type="button"
  >
    <span
      className="MuiButton-label-5"
    >
      Cancel
    </span>
    <span
      className="MuiTouchRipple-root-21"
    />
  </button>
  <button
    className="MuiButtonBase-root-19 MuiButtonBase-disabled-20 MuiButton-root-4 MuiButton-raised-9 MuiButton-raisedPrimary-10 MuiButton-disabled-13"
    disabled={true}
    onBlur={[Function]}
    onClick={[MockFunction]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex="-1"
    type="button"
  >
    <span
      className="MuiButton-label-5"
    >
      Save
    </span>
  </button>
</div>
`;
